[WPE][GTK] Propagate the font's computed locale to HarfBuzz

mbrodesser-Igalia opened the pull request:
#### c30953f1f00cf190463b4f21b0d9f1fef76b0591
<pre>
[WPE][GTK] Propagate the font&apos;s computed locale to HarfBuzz
<a href="https://bugs.webkit.org/show_bug.cgi?id=287629">https://bugs.webkit.org/show_bug.cgi?id=287629</a>

Reviewed by Carlos Garcia Campos.

GTK and WPE use HarfBuzz and the language wasn&apos;t propagated to it.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/lang-attribute-affects-rendering-of-second-text-run-expected-mismatch.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/lang-attribute-affects-rendering-of-second-text-run-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/lang-attribute-affects-rendering-of-second-text-run.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/platform/graphics/skia/ComplexTextControllerSkia.cpp:
(WebCore::ComplexTextController::collectComplexTextRunsForCharacters):

Canonical link: <a href="https://commits.webkit.org/291691@main">https://commits.webkit.org/291691@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eff95655af30871e43dbb0c33d3bfbe6e656447b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/93711 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/13284 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/3020 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/98717 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/44237 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/95761 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/13580 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/21727 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/71539 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/28910 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/96713 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/10114 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/84695 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/51873 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/9788 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/2322 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/43552 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/80061 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/2390 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/100751 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/20763 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/15146 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/80554 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/21015 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/80637 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/79892 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/19874 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/24440 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/1793 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/13916 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/20747 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/25925 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/20434 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/23894 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/22175 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on LayoutTests/platform/glib/TestExpectations:
```diff
@@ -3237,6 +3236,9 @@ webkit.org/b/68512 accessibility/aria-hidden-updates-alldescendants.html [ Failu
 webkit.org/b/77568 fast/text/locale-shaping-complex.html [ ImageOnlyFailure ]
 webkit.org/b/77568 fast/text/locale-shaping.html [ ImageOnlyFailure ]
 
+# font-feature-settings aren't implemented for GTK and WPE:
```
Are you sure it's not implemented?

carlosgcampos replied:
It could be a bug, no need to investigate further, just remove the comment.

khaledhosny commented on Source/WebCore/platform/graphics/skia/ComplexTextControllerSkia.cpp:
```diff
@@ -191,6 +192,14 @@ void ComplexTextController::collectComplexTextRunsForCharacters(std::span<const
+    // According to <https://datatracker.ietf.org/doc/html/rfc5646#section-2.1>
+    // "the language tags described in this document are sequences of characters
+    // from the US-ASCII [ISO646] repertoire.".
+    hb_buffer_set_language(buffer.get(), hb_language_from_string(m_font.fontDescription().computedLocale().string().ascii().data(), -1));
```
`hb_buffer_reset()` is called in the loop below, which will reset the language set on the buffer, so the language will be applied to the first run only.

carlosgcampos replied:
Good catch! The language should be set inside the loop

khaledhosny replied:
I don’t know how webkit segements the runs, but from that loop I assume at least multi-script string would be split into multiple runs, so a string like `fi αβ fi` would result in three runs.

mrego replied:
Why not adding the test directly as part of this patch here in WebKit? You can mark the PR at WPT as exported from WebKit so once the patch is approved here, it's automatically approved there too.

carlosgcampos commented on Source/WebCore/platform/graphics/skia/ComplexTextControllerSkia.cpp:
```diff
@@ -191,7 +192,15 @@ void ComplexTextController::collectComplexTextRunsForCharacters(std::span<const
+        // According to <https://datatracker.ietf.org/doc/html/rfc5646#section-2.1>
+        // "the language tags described in this document are sequences of characters
+        // from the US-ASCII [ISO646] repertoire.".
+        hb_buffer_set_language(buffer.get(), hb_language_from_string(m_font.fontDescription().computedLocale().string().ascii().data(), -1));
```
The language is the same for all runs, so I would get the hb_language_t once before the loop and use that value in every loop iteration.

Commit:
[WPE][GTK] Propagate the font's computed locale to HarfBuzz
https://bugs.webkit.org/show_bug.cgi?id=287629

Reviewed by Carlos Garcia Campos.

GTK and WPE use HarfBuzz and the language wasn't propagated to it.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/lang-attribute-affects-rendering-of-second-text-run-expected-mismatch.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/lang-attribute-affects-rendering-of-second-text-run-ref.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/lang-attribute-affects-rendering-of-second-text-run.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac/TestExpectations:
* Source/WebCore/platform/graphics/skia/ComplexTextControllerSkia.cpp:
(WebCore::ComplexTextController::collectComplexTextRunsForCharacters):

Canonical link: https://commits.webkit.org/291691@main
